### DISCUSSION TITLE
Paginated relational data properties

### DISCUSSION THREAD

**Discussion by mjanssen** [SeaQL/sea-orm#1232] [Q&A]
Hi,

I have been trying to return paginated relational data, but it seems to return a `Vec<Model, Option<Model>>` from the `find_also_related` method. What would be a good way to add properties to every entry, filled with data from the user entry?

<img width="623" alt="image" src="https://user-images.githubusercontent.com/1145427/202211984-0cc39765-dcbc-440d-b3ad-6633b0925727.png">

The code to request the data is as following:

```
() -> Result<(Vec<(Model, Option<super::user::Model>)>, u64, u64, usize), DbErr>

    let mut q = self::Entity::find().find_also_related(super::user::Entity);
    let paginator = q.order_by_desc(self::Column::CreatedAt).paginate(db, limit);

    let num_pages = paginator.num_pages().await?;
    let total_amount = paginator.num_items().await?;

    paginator
        .fetch_page(page - 1)
        .await
        .map(|e| (e.clone(), num_pages, total_amount, e.len()))
```

**Comment by billy1624** [comment]
Hey @mjanssen, welcome!! In this case I strongly recommend you to checkout [seaography](https://github.com/SeaQL/seaography), a GraphQL framework built on top of SeaORM which support relational select in batch.

**Comment by mjanssen** [comment]
Hi @billy1624, thanks for your answer! I didn't want to move to a GraphQL implementation, since the client application only uses regular REST calls to other API's. So implementing a GraphQL lib would be a bit overkill.

I have tried implementing a user relationship by setting `pub user: Option<super::user::Entity>` on my original schema, but this is not accepted.

Is there no other option than implementing GraphQL?

**Comment by mjanssen** [comment]
Managed to get the desired result using a sepperate for loop in the `.map` method.

```
paginator.fetch_page(page - 1).await.map(|items| {
    let mut result: Vec<AppEventItem> = Vec::new();
    for item in items {
        let (event, user) = item;
        let res = AppEventItem {
            id: event.id,
            created_at: event.created_at,
            description: event.description,
            event_type: event.event_type,
            target_id: event.target_id,
            user_id: event.user_id,
            user,
        };

        result.push(res);
    }

    return (result.clone(), num_pages, total_amount, result.len());
})
```

But feels like a hacky solution 🤔

**Comment by mjanssen** [comment]
Managed to fix the result by using custom `.column_as()` in combination with `.join()`. Eventually using `.into_model::<>()` with a custom resultset to manage the response. This returns a `Vec<>` with the desired response.

```
let mut q = self::Entity::find()
    .column_as(super::user::Column::Name, "user_name")
    .column_as(super::target::Column::Name, "target_name")
    .column_as(self::Column::EventType, "event_type")
    .join(JoinType::InnerJoin, self::Relation::User.def())
    .join(JoinType::InnerJoin, self::Relation::Target.def());
```

The only downside using this implementation is that you cannot have the entire `User` model implemented (unless you map all these fields separately).
